src/storage/src/render/multi_worker_persist_sink.rs

```diff
@@ -0,0 +1,1123 @@
+                shard_id,
+                &format!("compute::persist_sink::write_batches {}", collection_id),
+                Arc::new(target_relation_desc),
+                Arc::new(UnitSchema),
+            )
+            .await
+            .expect("could not open persist shard");
+
+        // The current input frontiers.
+        let mut batch_descriptions_frontier = Antichain::from_elem(TimelyTimestamp::minimum());
+        let mut desired_frontier = Antichain::from_elem(TimelyTimestamp::minimum());
+
+        // The frontiers of the inputs we have processed, used to avoid redoing work
+        let mut processed_desired_frontier = Antichain::from_elem(TimelyTimestamp::minimum());
+        let mut processed_descriptions_frontier = Antichain::from_elem(TimelyTimestamp::minimum());
+
+        // A "safe" choice for the lower of new batches we are creating.
+        let mut operator_batch_lower = Antichain::from_elem(TimelyTimestamp::minimum());
+
+        loop {
+            tokio::select! {
+                Some(event) = descriptions_input.next_mut() => {
+                    match event {
+                        Event::Data(cap, data) => {
+                            // Ingest new batch descriptions.
+                            for description in data.drain(..) {
+                                if collection_id.is_user() {
+                                    trace!(
+                                        "persist_sink {collection_id}/{shard_id}: \
+                                            write_batches: \
+                                            new_description: {:?}, \
+                                            desired_frontier: {:?}, \
+                                            batch_descriptions_frontier: {:?}",
+                                        description,
+                                        desired_frontier,
+                                        batch_descriptions_frontier,
+                                    );
+                                }
+                                let existing = in_flight_batches.insert(
+                                    description.clone(),
+                                    cap.delayed(description.0.first().unwrap()),
+                                );
+                                assert!(
+                                    existing.is_none(),
+                                    "write_batches: sink {} got more than one \
+                                        batch for description {:?}, in-flight: {:?}",
+                                    collection_id,
+                                    description,
+                                    in_flight_batches
+                                );
+                            }
+
+                            continue;
+                        }
+                        Event::Progress(frontier) => {
+                            batch_descriptions_frontier = frontier;
```

Comment:
Why do we care about this frontier? Batch descriptions are guaranteed to not be overlapping and we can process them immediately

Reply:
We use this as a light optimization on avoiding work after each `select!` to avoid emitting new batches if this or the desired frontier hasn't moved!